### PR TITLE
[java] InefficientEmptyStringCheckRule now reports String.trim().isEmpty()

### PR DESCRIPTION
**Pull request by Oowekyala**
**Issue solved** #290 
**PR Description:** InefficientEmptyStringCheckRule now reports calls to String.trim().isEmpty(), as in the following snippet:

```java
final String str = "   test.  ";
// the next line should cause a violation of the InefficientEmptyStringCheck but does not
if (str.trim().isEmpty()) {
    System.out.println("String is empty");
}
```

### CODE REVIEW

File: pmd-java/src/test/java/net/sourceforge/pmd/lang/java/rule/strings/InefficientEmptyStringCheckTest.java [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+
+package net.sourceforge.pmd.lang.java.rule.strings;
+
+import net.sourceforge.pmd.testframework.SimpleAggregatorTst;
+
+/**
+ * @author Clément Fournier (clement.fournier@insa-rennes.fr)
+ *
+ */
+public class InefficientEmptyStringCheckTest extends SimpleAggregatorTst {
```
Comment (jsotuyod) [inline pmd-java/src/test/java/net/sourceforge/pmd/lang/java/rule/strings/InefficientEmptyStringCheckTest.java]: This class should not be necessary, the [`StringRulesTest`](https://github.com/pmd/pmd/blob/master/pmd-java/src/test/java/net/sourceforge/pmd/lang/java/rule/strings/StringsRulesTest.java#L20) already checks the tests on that xml spec.

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/strings/InefficientEmptyStringCheckRule.java [hunk 2]
```diff
@@ -5,47 +5,63 @@
...
+        
+        if (node.jjtGetNumChildren() > 3) {
+            // Check last suffix
+            if (!"isEmpty".equals(node.jjtGetChild(node.jjtGetNumChildren() - 2).getImage())) {
+                return data;
+            }
+            
+            Node prevCall = node.jjtGetChild(node.jjtGetNumChildren() - 4);
+            String target = prevCall.jjtGetNumChildren() > 0 ? prevCall.jjtGetChild(0).getImage() : prevCall.getImage();
+            if (target != null && target.indexOf("trim") != -1) {
```
Comment (jsotuyod) [inline pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/strings/InefficientEmptyStringCheckRule.java]: any reason for not doing `target != null && "trim".equals(target)`?

Comment (Oowekyala) [inline pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/strings/InefficientEmptyStringCheckRule.java]: In the following statement, considering a String `foo` :
```java
boolean b = foo.trim().isEmpty();
```
The expression's AST is of the form 
```
PrimaryExpression
    PrimaryPrefix
        Name
    PrimarySuffix
        Arguments
    PrimarySuffix
    PrimarySuffix
        Arguments
```
Where the image of `Name` (`target`) actually is `foo.trim`, so we cannot use equals. Perhaps we could do `if (target != null && target.indexOf(".trim") != -1) ` (with a dot) so as to not match other methods

File: pmd-java/src/test/java/net/sourceforge/pmd/lang/java/rule/strings/InefficientEmptyStringCheckTest.java (hunk 1 above)
Comment (Oowekyala) [inline pmd-java/src/test/java/net/sourceforge/pmd/lang/java/rule/strings/InefficientEmptyStringCheckTest.java]: Oh ok I can delete it I just used it to test the rule individually

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/strings/InefficientEmptyStringCheckRule.java (hunk 2 above)
Comment (jsotuyod) [inline pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/strings/InefficientEmptyStringCheckRule.java]: I guessed so. Yes, maybe even `target.endsWith(".trim")` to be more restrictive

Comment (Oowekyala) [inline pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/strings/InefficientEmptyStringCheckRule.java]: Well after verification we cannot do that as the image of the call to `trim` in a chain call such as `foo.toString().trim().isEmpty()` is `trim` and not `.trim`...

Comment (jsotuyod) [inline pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/strings/InefficientEmptyStringCheckRule.java]: @Oowekyala that's a great idea!

Comment (Oowekyala) [inline pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/strings/InefficientEmptyStringCheckRule.java]: Done !

**Comment by jsotuyod** [comment]
Thanks for the contribution! This will be part of the next cycle of PMD releases!

**Pull request merged**